File: common/src/main/java/org/tdf/common/serialize/Codec.java
package org.tdf.common.serialize;

import java.util.function.Function;

/**
 * encoder and decoder from a type to another type
 *
 * @param <K> type
 */
public interface Codec<K> {
    static Codec<byte[]> identity() {
        return (Codec<byte[]>) Codecs.IDENTITY;
    }

    static <K> Codec<K> newInstance(
        Function<? super K, byte[]> encoder,
        Function<byte[], ? extends K> decoder
    ) {
        return new Codec<K>() {
            @Override
            public Function<? super K, byte[]> getEncoder() {
                return encoder;
            }

            @Override
            public Function<byte[], ? extends K> getDecoder() {
                return decoder;
            }
        };
    }

    Function<? super K, byte[]> getEncoder();

    Function<byte[], ? extends K> getDecoder();
}

File: common/src/main/java/org/tdf/common/serialize/Codec.kt
package org.tdf.common.serialize

import java.util.function.Function

/**
 * encoder and decoder from a type to another type
 *
 * @param <K> type
</K> */
interface Codec<K> {
    val encoder: Function<in K, ByteArray>
    val decoder: Function<ByteArray, out K>

    companion object {
        @JvmStatic
        fun <K> identity(): Codec<K> {
            return Codecs.identity as Codec<K>
        }

        fun <K> create(
            encoder: Function<in K, ByteArray>,
            decoder: Function<ByteArray, out K>
        ): Codec<K> {
            return CodecImpl(encoder, decoder)
        }
    }
}

internal class CodecImpl<K>(
    override val encoder: Function<in K, ByteArray>,
    override val decoder: Function<ByteArray, out K>
) : Codec<K>